### PR TITLE
[Issue 877]Fix send stuck

### LINKED ISSUE #877: partitionProducer Send stuck
We use fork version in this line：https://github.com/billowqiu/pulsar-client-go/blob/1dc06ae37025b011ecc345312a65d797b436dedb/pulsar/producer_partition.go#L745

 but the upstream version have the same problem.

#### Expected behavior

partitionProducer.Send should return after success or timeout.

#### Actual behavior

stuck in doneCh.

image

image

#### Steps to reproduce

How can we reproduce the issue

#### System configuration
**Pulsar version**: x.y

### PR DESCRIPTION
**Pull request by billowqiu**
Fixes #877

### Motivation

Use select to avoid channel stuck. 

### Modifications

- like consumer.Receive func,  use for...select idiom.

### CODE REVIEW

**Comment by Gleiphir2769** [comment]
Hi, could you upload the full error stack? I think we need to find out real reason for doneCh can not be closed. Because doneCh is never closed means that the callback of `sendRequest` is not be called. This shouldn't happen.

**Comment by billowqiu** [comment]
> Hi, could you upload the full error stack? I think we need to find out the real reason for doneCh can not be closed. Because doneCh is never closed means that the callback of `sendRequest` is not be called. This shouldn't happen.

In the issue link have the stack. I think wait in a channel should with select timeout context to avoid stuck.

**Comment by Gleiphir2769** [comment]
> In the issue link have the stack. I think wait in a channel should with select timeout context to avoid stuck.

You are right, timeout context can avoid `doneCh` stuck. 

But `send()` stuck shouldn't happen. I think add select just skip this problem.

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -1035,9 +1035,15 @@ func (p *partitionProducer) Send(ctx context.Context, msg *ProducerMessage) (Mes
 	}, true)
 
 	// wait for send request to finish
-	<-doneCh
-
-	return msgID, err
+	for {
```
Comment (Gleiphir2769) [inline pulsar/producer_partition.go:1038]:
```suggestion
```
Hi, I don't think a loop is needed here.

**Review by Gleiphir2769 — CHANGES_REQUESTED** [review]

File: pulsar/producer_partition.go (hunk 1 above)
Comment (billowqiu) [reply pulsar/producer_partition.go:1038]: I also think this loop is useless.

File: pulsar/producer_partition.go [hunk 2]
```diff
@@ -1035,9 +1035,13 @@ func (p *partitionProducer) Send(ctx context.Context, msg *ProducerMessage) (Mes
...
 	// wait for send request to finish
-	<-doneCh
-
-	return msgID, err
+	select {
+	case <-ctx.Done():
+		isDone.Store(true)
+		return msgID, ctx.Err()
+	case <-doneCh:
+		return msgID, err
```
Comment (RobertIndie) [inline pulsar/producer_partition.go:1043]: We shouldn't return msgID and err here. They are not assigned here.

**Review by RobertIndie — CHANGES_REQUESTED** [review]

**Comment by RobertIndie** [comment]
Fixed by https://github.com/apache/pulsar-client-go/pull/1053

**Pull request closed without merging**